### PR TITLE
Allow to use complex types in rules

### PR DESCRIPTION
**Pull request by ISkomorokh**
## Definition of the problem
Quite often there is a necessity to release features based on the product version. Let's say we want to enable some feature for all the clients which are using our product and it's version is greater than 10.7.0. In order to archive that we have to do the following:

1. Introduce attributes which will represent product version like this: `{ majorVersion: 10, minorVersion: 7, miniVersion: 0 }` (this part is easy and should be done only once)
2. Create a rule for the feature like this:
image

This becomes a nightmare to setup and maintain when you have tens of features and you have to target multiple versions.

## The idea how this could be simplified
1. Introduce an attribute like this: `{ productVersion: "10.7.0" }`
2. Create a simple rule like this:
image

## Implementation
In this PR I've completely preserved original logic. I've just used `objects` instead of types from Json.NET lib and refactored few things a bit. The idea is to use implementations of `IEqualityComparer`, `IComparer` etc. and also to convert objects from one type to another (mostly from strings to complex types).

In order to convert from `string` to `System.Version` type, consumers of LD client should provide an implementation of `ITypeConverter` and register it in `Configuration` which is being provided to LD client.

I was thinking about using capabilities of `System.ComponentModel.TypeConverter` which is perfect for this task, but it has no implementation on `.net core 1.0` platform. That's why I've added `ITypeConverter` and it seems to be more explicit :) Anyway, this part could be `#if def`.

I've tried to cover everything with tests.

So this table will be extended pretty much as this PR will make it possible to compare not only primitives
image

### Breaking changes

1. User.Custom - type was changed from `IDictionary<string, JToken>` to `IDictionary<string,object>`. As for me it was a bit "leaky", because we were adding stings, ints etc into custom attributes, but they were exposed as "Json" types.
2. User's extension methods, namely overloads of `AndCustomAttribute` were deleted. Only one that accepts `object` was left.
These changes also removing direct dependency on Newtonsoft.Json lib (basically during runtime it is possible to use binding redirects when version used by clients differs, but things are not that good during compile time).

Hope this will make us a bit close to enterprise level requirements.

@drichelson , @jkodumal  please, review

### CODE REVIEW

**Comment by jkodumal** [comment]
Thanks @ISkomorokh -- we're starting to review now.

File: src/LaunchDarkly.Client/Operators/Default/GreatedThan.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+﻿using System;
+
+namespace LaunchDarkly.Client.Operators
+{
+
+    internal class GreatedThan : IOperatorExecutor
```
Comment (kbaranovska) [inline src/LaunchDarkly.Client/Operators/Default/GreatedThan.cs]: Probably, name should be Greate**r**

File: src/LaunchDarkly.Client/Clause.cs [hunk 2]
```diff
@@ -1,84 +1,67 @@
...
+using Newtonsoft.Json;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace LaunchDarkly.Client {
+    internal class Clause
+    {
+
+        internal string Attribute { get; }
+        internal string Op { get; }
```
Comment (kbaranovska) [inline src/LaunchDarkly.Client/Clause.cs]: What does the Op mean? How about renaming it to more obvious name?

File: src/LaunchDarkly.Client/Operators/Default/GreatedThan.cs (hunk 1 above)
Comment (ISkomorokh) [inline src/LaunchDarkly.Client/Operators/Default/GreatedThan.cs]: Yeepp... Thanks. Fixed

File: src/LaunchDarkly.Client/Clause.cs (hunk 2 above)
Comment (ISkomorokh) [inline src/LaunchDarkly.Client/Clause.cs]: It means Operation. Left original naming in order to keep PR clear. Will be revised in future PRs. Thanks

**Comment by ISkomorokh** [comment]
@jkodumal any news?

**Comment by arun251** [comment]
Hi @ISkomorokh , this is a very interesting idea. We strive for our feature rules to be platform-independent and consistent across all SDKs. If we accepted this PR, we would need to update our application and all of our SDKs to support new types and operators. This specific case isn't high on our backlog at this time, but if it is, we will take a look at your PR.

As a workaround, I would suggest converting the version to a number or numeric string. For instance, `10.7.0` could be represented as the number `100700` or the string `"0100700"` or even `"10.07.00"`, and then the standard operators work fairly well. They're not as clear as `10.7.0`, but they are well supported.

**Comment by ISkomorokh** [comment]
sigh... representing versions that way doesn't wok wall and it is unreal to maintain especially when managers come into play.  https://softwareengineering.stackexchange.com/questions/313748/convert-version-string-to-integer/313750#313750

**Pull request closed without merging**